### PR TITLE
Remove `hpc_save`

### PR DESCRIPTION
**Pull request by jjenniferdai**
## What does this PR do?

Item 2. in https://github.com/PyTorchLightning/pytorch-lightning/issues/9407#issuecomment-991399437
Part of #9407 cleanup, remove `hpc_save`.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 1]
```diff
@@ -299,39 +299,11 @@ def restore_lr_schedulers(self) -> None:
     # PRIVATE OPS
     # ----------------------------------
 
-    def hpc_save(self, folderpath: str, logger: Optional[LightningLoggerBase]) -> str:
-        # make sure the checkpoint folder exists
+    def hpc_save_path(self, folderpath: str) -> str:
```
Comment (ananthsub) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:302]: super nit: you could make this, `max_ckpt_version_in_folder`, and `get_max_ckpt_path_from_folder` staticmethods. that's usually a sign that this class can be further streamlined by moving those out to utility functions. that could be for a separate issue/PR

File: pytorch_lightning/trainer/connectors/signal_connector.py [hunk 2]
```diff
@@ -61,7 +61,11 @@ def slurm_sigusr1_handler_fn(self, signum: Signals, frame: FrameType) -> None:
         if self.trainer.is_global_zero:
             # save weights
             log.info("handling SIGUSR1")
-            self.trainer.checkpoint_connector.hpc_save(self.trainer.weights_save_path, self.trainer.logger)
+            # save logger to make sure we get all the metrics
+            if self.trainer.logger:
+                self.trainer.logger.finalize("finished")
+            hpc_save_path = self.trainer.checkpoint_connector.hpc_save_path(self.trainer.weights_save_path)
+            self.trainer.save_checkpoint(hpc_save_path)
```
Comment (ananthsub) [inline pytorch_lightning/trainer/connectors/signal_connector.py:68]: this can't be under a `self.trainer.is_global_zero` check from L61. `trainer.save_checkpoint` relies on collectives, meaning it must be called on all ranks simultaneously. otherwise the user will see hangs in a distributed environment. 

To minimize changes in this PR, you could encapsulate just the logger finalization call under the `is_global_zero` check. `trainer.save_checkpoint` already takes care of all the other rank-specific checks. 

it looks like we don't have a test with deepspeed-3 and SLURM signal handling. in the code today, this would crash, because we need to save on all ranks in that case. 

so this is an awesome unification you're doing so we don't have to worry about divergent code paths like this anymore 😄

**Review by ananthsub — CHANGES_REQUESTED** [review]

File: pytorch_lightning/trainer/connectors/signal_connector.py (hunk 2 above)
Comment (ananthsub) [reply pytorch_lightning/trainer/connectors/signal_connector.py:68]: looking at what's done in the main trainer, the logger finalization is called on all ranks, and it's up to logger implementations to determine if they only want to be called on rank 0 or not.

https://github.com/PyTorchLightning/pytorch-lightning/blob/7ed3dbf1913351868835ac3c549a0fbf6c612e09/pytorch_lightning/trainer/trainer.py#L1429-L1446

So I don't think we need to guard this with the global zero check either.

Comment (jjenniferdai) [reply pytorch_lightning/trainer/connectors/signal_connector.py]: I think* the second finalize call can be removed, but someone correct me if I'm missing something :)

Comment (tchaton) [reply pytorch_lightning/trainer/connectors/signal_connector.py]: Hey @ananthsub. Great catch ! Yes, this should be covering only the requeing.

Yes, we should be call the logger finalize on all ranks, so let's remove the second one.

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 3]
```diff
@@ -489,3 +422,54 @@ def _get_loops_state_dict(self) -> Dict[str, Any]:
             "test_loop": self.trainer.test_loop.state_dict(),
             "predict_loop": self.trainer.predict_loop.state_dict(),
         }
+
+    # ----------------------------------
+    # UTILITIES
+    # ----------------------------------
+
+    @staticmethod
+    def max_ckpt_version_in_folder(dir_path: _PATH, name_key: str = "ckpt_") -> Optional[int]:
```
[the remark this answers is not in the document]
Comment (jjenniferdai) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:431]: protected? they're sort of like utility functions, `hpc_save_path()` gets called in `slurm_sigusr1_handler_fn`

File: pytorch_lightning/trainer/connectors/signal_connector.py [hunk 4]
```diff
@@ -61,8 +61,14 @@ def slurm_sigusr1_handler_fn(self, signum: Signals, frame: FrameType) -> None:
         if self.trainer.is_global_zero:
             # save weights
             log.info("handling SIGUSR1")
```
Comment (ananthsub) [inline pytorch_lightning/trainer/connectors/signal_connector.py:67]: super nit
```suggestion
        rank_zero_info("handling SIGUSR1")
```

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ananthsub) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:434]: @jjenniferdai you could make the utilities which are only called within this class private

File: pytorch_lightning/trainer/connectors/signal_connector.py [hunk 5]
```diff
@@ -65,8 +65,14 @@ def slurm_sigusr1_handler_fn(self, signum: _SIGNUM, frame: FrameType) -> None:
...
-            self.trainer.checkpoint_connector.hpc_save(self.trainer.weights_save_path, self.trainer.logger)
 
+        # save logger to make sure we get all the metrics
+        if self.trainer.logger:
+            self.trainer.logger.finalize("finished")
+        hpc_save_path = self.trainer.checkpoint_connector.hpc_save_path(self.trainer.weights_save_path)
+        self.trainer.save_checkpoint(hpc_save_path)
```
Comment (ananthsub) [inline pytorch_lightning/trainer/connectors/signal_connector.py:73]: this is so much cleaner!

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jjenniferdai) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:434]: some tests calling them though, should we just update those with e.g. `_CheckpointConnector__ max_ckpt_version_in_folder` ?

**Review by justusschock — APPROVED** [review]

**Review by ananthsub — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]

**Review by carmocca — APPROVED** [review]

**Review by Borda — APPROVED** [review]
